File: 2023-12-27/BT_2.py
n, m = map(int, input().split())
tteokList = sorted(list(map(int, input().split())))

maxHeight = 0
low, high = 0, tteokList[-1]

while low < high:
    cutLen = 0
    mid = (low + high) // 2

    for tLen in tteokList:
        cutLen += tLen - mid if tLen > mid else 0
    
    if m == cutLen:
        maxHeight = mid
        break
    elif m < cutLen:
        low = mid + 1
        maxHeight = max(maxHeight, mid)
    else:
        high = mid
print(maxHeight)

File: 2023-12-27/BT_2.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
const [L, datas] = fs.readFileSync(filePath).toString().trim().split('\n');

const solution = ([n, m], tteokList) => {
  tteokList.sort((a, b) => a - b);
  let maxHeight = 0;
  let [low, high] = [0, tteokList[n - 1]];

  while (low < high) {
    let mid = ~~((low + high) / 2);
    let cutLen = tteokList.reduce((sum, tLen) => (mid < tLen ? sum + tLen - mid : sum), 0);

    if (m === cutLen) {
      maxHeight = mid;
      break;
    } else if (m < cutLen) {
      low = mid + 1;
      maxHeight = Math.max(maxHeight, mid);
    } else {
      high = mid;
    }
  }
  console.log(maxHeight);
};

solution(L.split(' ').map(Number), datas.split(' ').map(Number));
